### Import Packages

In [9]:
# geospatial analysis
import geopandas as gpd
import restapi # downloading data
from scipy.spatial import cKDTree
from pyproj import Proj

# working with data
import numpy as np
import h5pyd
import pandas as pd
import xarray as xr

# web scraping
import requests
from bs4 import BeautifulSoup
import time
import re

# API
from gridstatusio import GridStatusClient

# working with files
import os

# work with time
import pytz
import dateutil
from datetime import timezone

# API key
GRIDSTATUS_API_KEY = os.environ["GRIDSTATUS_API_KEY"]

## Import Common Geometries

In [10]:
# California counties
us_counties = gpd.read_file("Data/Boundaries/cb_2018_us_county_500k/cb_2018_us_county_500k.shp")
ca_counties = us_counties[us_counties["STATEFP"] == "06"]
ca_counties.to_crs("EPSG:5070", inplace=True)

# California HUC8 subbasins
huc8_ca = gpd.read_file("Data/Boundaries/ca_HUC8.geojson")

/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/geopandas/geodataframe.py:1819: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  super().__setitem__(key, value)


## Electric Grid Headroom

### Downloading SCE

Other relevant links for Southern California Edison:

https://drpep-sce2.opendata.arcgis.com/maps/f76174d459a14545b2b13110d1cc6fba/about

Tutorial: https://drpep.sce.com/drpep/cdn/19/drpep-interactive-user-guide/index.html#/

In [2]:
# download Southern California Edison data
# sce_url = "https://drpep.sce.com/arcgis_server/rest/services/Hosted/AVL_LOAD_CAP_TOGGLE/FeatureServer/1"
sce_url = "https://drpep.sce.com/arcgis_server/rest/services/Hosted/AVL_LOAD_CAP_TOGGLE/FeatureServer/0"
sce_layer = restapi.FeatureLayer(sce_url)
# sce_layer.export_layer("Data/Hosting_Capacity/sce.geojson")

/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/restapi/rest_utils.py:42: UserWarning: no request client has been set, using default client
  warnings.warn('no request client has been set, using default client')


In [34]:
sce_layer.list_fields()

['circuit_name',
 'substation_name',
 'date_last_update',
 'plan_year',
 'system_name',
 'a_bank',
 'alc_year1',
 'alc_year2',
 'alc_year3',
 'alc_year4',
 'alc_year5',
 'subst_alc_year1',
 'subst_alc_year2',
 'subst_alc_year3',
 'subst_alc_year4',
 'subst_alc_year5',
 'a_bank_alc_year1',
 'a_bank_alc_year2',
 'a_bank_alc_year3',
 'a_bank_alc_year4',
 'a_bank_alc_year5',
 'circuit_voltage',
 'org_alc_year1',
 'org_alc_year2',
 'org_alc_year3',
 'org_alc_year4',
 'org_alc_year5',
 'ckt_capacity_y1',
 'ckt_capacity_y2',
 'ckt_capacity_y3',
 'ckt_capacity_y4',
 'ckt_capacity_y5',
 'subst_capacity_y1',
 'subst_capacity_y2',
 'subst_capacity_y3',
 'subst_capacity_y4',
 'subst_capacity_y5',
 'subst_org_alc_year1',
 'subst_org_alc_year2',
 'subst_org_alc_year3',
 'subst_org_alc_year4',
 'subst_org_alc_year5',
 'subtrans',
 'subtrans_alc_year1',
 'subtrans_alc_year2',
 'subtrans_alc_year3',
 'subtrans_alc_year4',
 'subtrans_alc_year5',
 'y1ckt_f',
 'y2ckt_f',
 'y3ckt_f',
 'y4ckt_f',
 'y5ckt_f'

In [28]:
cursor = restapi.SearchCursor(sce_layer, ["objectid"])

In [ ]:
# verify SCE data downloaded properly
sce_alc = gpd.read_file("Data/Hosting_Capacity/SCE/Available Load Capacity.geojson")

# for simplicity, let's say year 1 (2024)
sce_alc.drop(columns=sum([[f"alc_year{i}", f"a_bank_alc_year{i}", f"org_alc_year{i}", f"ckt_capacity_y{i}", f"subst_capacity_y{i}", f"subst_org_alc_year{i}", f"subtrans_alc_year{i}", f"y{i}ckt_f", f"y{i}sub_f", f"y{i}subt_f", f"y{i}abank_f", f"y{i}ckt_sub_f", f"y{i}ckt_subt_f", f"y{i}ckt_abank_f", f"y{i}sub_subt_f", f"y{i}sub_abank_f", f"y{i}subt_abank_f", f"y{i}ckt_sub_subt_f", f"y{i}ckt_sub_abank_f", f"y{i}ckt_subt_abank_f", f"y{i}sub_subt_abank_f", f"y{i}ckt_sub_subt_abank_f"] for i in range(2,6)], []), inplace=True)

# sce_alc.to_file("Data/SCE_available_load_capacity.geojson", driver="GeoJSON") # somehow the file size becomes bigger after this

### Downloading PGE

In [20]:
# verify PG&E data download properly
pge_ica = gpd.read_file("Data/Hosting_Capacity/PG&E/ICA_Results/LineDetail.shp")

### Downloading SDG&E

In [2]:
# verify SDG&E data download properly
sdge_ica = gpd.read_file("Data/Hosting_Capacity/SDG&E/ICA_MAP_PROD_LoadCapacityGrids_VW_-864651790911740139.geojson")

In [3]:
sdge_ica

OBJECTID JOIN_COUNT TARGET_FID JOIN_FID SHAPE_LENG  VOLTAGE  \
0       28922611       None       None     None       None       12   
1       28922612       None       None     None       None       12   
2       28922613       None       None     None       None       12   
3       28922614       None       None     None       None       12   
4       28922615       None       None     None       None       12   
...          ...        ...        ...      ...        ...      ...   
477050  29399661       None       None     None       None       12   
477051  29399662       None       None     None       None       12   
477052  29399663       None       None     None       None       12   
477053  29399664       None       None     None       None       12   
477054  29399665       None       None     None       None       12   

        LINE_SEGMENT_NUMBER  ICAWOF_UNIGENERATION  ICAWOF_UNILOAD  \
0                    141921                   0.0             0.8   
1                    141921                   0.0             0.8   
2                    141921                   0.0             0.8   
3                    139746                   0.0             0.0   
4                    139746                   0.0             0.0   
...                     ...                   ...             ...   
477050               177379                   0.0             0.2   
477051               177379                   0.0             0.2   
477052               179642                   0.0             0.2   
477053               179642                   0.0             0.2   
477054               179642                   0.0             0.2   

        ICAWOF_PVGENERATION OHUG CIRCUIT_NAME LABELTEXT  \
0                       0.2   OH          456      None   
1                       0.2   OH          456      None   
2                       0.2   OH          456      None   
3                       0.2   OH          456      None   
4                       0.2   OH          456      None   
...                     ...  ...          ...       ...   
477050                  0.1   OH          209      None   
477051                  0.1   OH          209      None   
477052                  0.1   OH          209      None   
477053                  0.1   OH          209      None   
477054                  0.1   OH          209      None   

        ICAWNOF_UNIGENERATION  ICAWNOF_UNILOAD  ICAWNOF_PVGENERATION  \
0                         0.0              0.8                   0.5   
1                         0.0              0.8                   0.5   
2                         0.0              0.8                   0.5   
3                         0.7              0.0                   0.9   
4                         0.7              0.0                   0.9   
...                       ...              ...                   ...   
477050                    0.0              0.2                   0.1   
477051                    0.0              0.2                   0.1   
477052                    0.0              0.2                   0.1   
477053                    0.0              0.2                   0.1   
477054                    0.0              0.2                   0.1   

       LABELTEXT_LCA RESTRICTED    SUBID  \
0         Up To 1.00          N      ASH   
1         Up To 1.00          N      ASH   
2         Up To 1.00          N      ASH   
3                  0          N      ASH   
4                  0          N      ASH   
...              ...        ...      ...   
477050    Up To 1.00          N  MELROSE   
477051    Up To 1.00          N  MELROSE   
477052    Up To 1.00          N  MELROSE   
477053    Up To 1.00          N  MELROSE   
477054    Up To 1.00          N  MELROSE   

                                                 geometry  
0       POLYGON ((-117.06446 33.13968, -117.06446 33.1...  
1       POLYGON ((-117.06527 33.13968, -117.06527 33.1...  
2       POLYGON ((-117.06527 33.13934, -117

## Electricity Cost

Prices taken from EnergySage. https://www.energysage.com/local-data/electricity-cost/ca/

In [3]:
county_regex = re.compile(rf"The average residential electricity rate in Alameda County, CA is (\d+) ¢/kWh")

matching = county_regex.findall("The average residential electricity rate in Alameda County, CA is 30 ¢/kWh")

for matches in matching:
    print(matches)

30


In [4]:
# get cost for each California county

ca_counties["Electricity Price (cents/kWh)"] = np.nan # initialize the column.

# find EnergySage URL
energy_sage_ca = "https://www.energysage.com/local-data/electricity-cost/ca/"

for idx, county in ca_counties.iterrows():
    print(county["NAME"])

    county_suffix = county["NAME"].replace(" ", "-").lower() + "-county" # convert county name to URL form

    county_energy_sage_url = energy_sage_ca + county_suffix

    county_energy_sage_html = requests.get(county_energy_sage_url) # retrieve website content

    # use regular expressions to extract electricity price
    county_regex = re.compile(rf"The average residential electricity rate in {county['NAME']} County, CA is (\d+) ¢/kWh")

    try:
        county_energy_sage_html = requests.get(county_energy_sage_url)
        matching = county_regex.findall(county_energy_sage_html.text)

        if len(matching) > 1:
            print(f"{county['NAME']} has more than one match.")
        elif len(matching) == 0:
            print(f"{county['NAME']} has no matches.")

        for matches in matching:
            ca_counties.loc[idx, 'Electricity Price (cents/kWh)'] = float(matches)

    except:
        print(county_suffix)

    time.sleep(1)


ca_counties["Electricity Price ($/MWh)"] = ca_counties["Electricity Price (cents/kWh)"]*10 # convert units (cents/kWh) to ($/MWh)

/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/geopandas/geodataframe.py:1819: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  super().__setitem__(key, value)


Alameda
Alpine
Alpine has no matches.
Amador
Contra Costa
Fresno
Humboldt
Lake
Los Angeles
Merced
Nevada
San Mateo
Santa Clara
Yuba
Mendocino
Mono
Mono has no matches.
Napa
Orange
San Benito
San Diego
San Luis Obispo
Santa Barbara
Sonoma
Sutter
Tulare
Yolo
Madera
Calaveras
Mariposa
Solano
Plumas
Siskiyou
Inyo
Monterey
Stanislaus
Sierra
Sierra has no matches.
Shasta
Colusa
Colusa has no matches.
Marin
San Joaquin
Kings
Lassen
Lassen has no matches.
San Francisco
Glenn
Glenn has no matches.
Imperial
Placer
Tuolumne
Ventura
Tehama
Modoc
Modoc has no matches.
Santa Cruz
Trinity
Trinity has no matches.
Kern
San Bernardino
El Dorado
Sacramento
Butte
Del Norte
Riverside


/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/geopandas/geodataframe.py:1819: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  super().__setitem__(key, value)


In [ ]:
ca_counties.to_file("Data/Boundaries/ca_county_prices_EnergySage.geojson", driver="GeoJSON")

Try again: Prices Taken from GridStatus.io. https://www.gridstatus.io/datasets/caiso_lmp_day_ahead_hourly

In [21]:
# data source: has data from 2019-03-21 PDT to present.
# let's do 2023 data. (don't do a leap year to avoid confusion about time indexing)

# takes around 1 minute to run

client = GridStatusClient(GRIDSTATUS_API_KEY)

df = client.get_dataset(
    dataset="caiso_lmp_day_ahead_hourly",
    start="2023-01-01 00:00:00",
    end="2023-12-31 01:00:00",
    filter_column="location_type",
    filter_value="AP Node",
    resample="1 year"
)

2025-09-04 23:14:21 - INFO - Fetching Page 1...
2025-09-04 23:14:21 - INFO - GET https://api.gridstatus.io/v1/datasets/caiso_lmp_day_ahead_hourly/query
2025-09-04 23:14:21 - INFO - Params: {'start_time': Timestamp('2023-01-01 00:00:00'), 'end_time': Timestamp('2023-12-31 01:00:00'), 'publish_time_start': None, 'publish_time_end': None, 'limit': None, 'page': 1, 'page_size': None, 'resample_frequency': '1 year', 'resample_by': None, 'resample_function': 'mean', 'publish_time': None, 'timezone': None, 'cursor': '', 'filter_column': 'location_type', 'filter_value': 'AP Node', 'filter_operator': '=', 'return_format': 'json', 'json_schema': 'array-of-arrays'}
2025-09-04 23:15:07 - INFO - Done in 45.64 seconds. 
2025-09-04 23:15:07 - INFO - Total number of rows: 2523


In [33]:
ca_counties.sort_values(by="NAME")

STATEFP COUNTYFP  COUNTYNS        AFFGEOID  GEOID             NAME LSAD  \
37        06      001  01675839  0500000US06001  06001          Alameda   06   
38        06      003  01675840  0500000US06003  06003           Alpine   06   
39        06      005  01675841  0500000US06005  06005           Amador   06   
3056      06      007  01675842  0500000US06007  06007            Butte   06   
752       06      009  01675885  0500000US06009  06009        Calaveras   06   
1188      06      011  01675902  0500000US06011  06011           Colusa   06   
40        06      013  01675903  0500000US06013  06013     Contra Costa   06   
3111      06      015  01682074  0500000US06015  06015        Del Norte   06   
3004      06      017  00277273  0500000US06017  06017        El Dorado   06   
41        06      019  00277274  0500000US06019  06019           Fresno   06   
1271      06      021  00277275  0500000US06021  06021            Glenn   06   
42        06      023  01681908  0500000US06023  06023         Humboldt   06   
1272      06      025  00277277  0500000US06025  06025         Imperial   06   
1023      06      027  01804637  0500000US06027  06027             Inyo   06   
2872      06      029  02054176  0500000US06029  06029             Kern   06   
1259      06      031  00277280  0500000US06031  06031            Kings   06   
43        06      033  00277281  0500000US06033  06033             Lake   06   
1260      06      035  01693324  0500000US06035  06035           Lassen   06   
44        06      037  00277283  0500000US06037  06037      Los Angeles   06   
735       06      039  00277284  0500000US06039  06039           Madera   06   
1212      06      041  00277285  0500000US06041  06041            Marin   06   
776       06      043  00277286  0500000US06043  06043         Mariposa   06   
337       06      045  00277287  0500000US06045  06045        Mendocino   06   
45        06      047  00277288  0500000US06047  06047           Merced   06   
2385      06      049  00277289  0500000US06049  06049            Modoc   06   
338       06      051  00277290  0500000US06051  06051             Mono   06   
1024      06      053  00277291  0500000US06053  06053         Monterey   06   
456       06      055  00277292  0500000US06055  06055             Napa   06   
46        06      057  01682927  0500000US06057  06057           Nevada   06   
457       06      059  00277294  0500000US06059  06059           Orange   06   
1273      06      061  00277295  0500000US06061  06061           Placer   06   
949       06      063  00277296  0500000US06063  06063           Plumas   06   
3215      06      065  00277297  0500000US06065  06065        Riverside   06   
3028      06      067  00277298  0500000US06067  06067       Sacramento   06   
458       06      069  00277299  0500000US06069  06069       San Benito   06   
2873      06      071  00277300  0500000US06071  06071   San Bernardino   06   
459       06      073  00277301  0500000US06073  06073        San Diego   06   
1261      06      075  00277302  0500000US06075  06075    San Francisco   06   
1213      06      077  00277303  0500000US06077  06077      San Joaquin   06   
460       06      079  00277304  0500000US06079  06079  San Luis Obispo   06   
47        06      081  00277305  0500000US06081  06081        San Mateo   06   
461       06      083  00277306  0500000US06083  06083    Santa Barbara   06   
48        06      085  00277307  0500000US06085  06085      Santa Clara   06   
2386      06      087  00277308  0500000US06087  06087       Santa Cruz   06   
1070      06      089  01682610  0500000US06089  06089           Shasta   06   
1038      06      091  00277310  0500000US06091  06091           Sierra   06   
955       06      093  00277311  0500000US06093  06093         Siskiyou   06   
777       06      095  00277312  0500000US06095  06095           Solano   06   
462       06      097  01657246  0500000US06097  06097           Sonoma   06   
1025 

In [31]:
df

interval_start_utc          interval_end_utc  \
0    2023-01-01 00:00:00+00:00 2024-01-01 00:00:00+00:00   
1    2023-01-01 00:00:00+00:00 2024-01-01 00:00:00+00:00   
2    2023-01-01 00:00:00+00:00 2024-01-01 00:00:00+00:00   
3    2023-01-01 00:00:00+00:00 2024-01-01 00:00:00+00:00   
4    2023-01-01 00:00:00+00:00 2024-01-01 00:00:00+00:00   
...                        ...                       ...   
2518 2023-01-01 00:00:00+00:00 2024-01-01 00:00:00+00:00   
2519 2023-01-01 00:00:00+00:00 2024-01-01 00:00:00+00:00   
2520 2023-01-01 00:00:00+00:00 2024-01-01 00:00:00+00:00   
2521 2023-01-01 00:00:00+00:00 2024-01-01 00:00:00+00:00   
2522 2023-01-01 00:00:00+00:00 2024-01-01 00:00:00+00:00   

                       location        lmp     energy  congestion      loss  
0           AFPR_1_TOT_GEN-APND  59.874700  62.360637    0.000000 -2.485937  
1          AFTON_3_AFTONCC-APND  59.645668  62.360637    0.000000 -2.714970  
2           AGRICO_6_PL3N5-APND  58.463239  62.360637   -3.030372 -0.867025  
3            AGRICO_7_UNIT-APND  58.463239  62.360637   -3.030372 -0.867025  
4           ALAMIT_2_PL1X3-APND  61.251450  62.360637   -0.474312 -0.634875  
...                         ...        ...        ...         ...       ...  
2518         WP_2_WPCC5MSG-APND  59.657701  62.360637    0.000000 -2.702936  
2519   WRW_2_BLUERDG_PLANT-APND  60.003069  62.360637    0.000000 -2.357568  
2520     WSH_1_WESTSIDEHYD-APND  60.010039  62.360637    0.000000 -2.350598  
2521  W TRACY_7_UNITS 8-10-APND  59.893395  62.360637    0.000000 -2.467242  
2522             YU_6_TGEN-APND  60.020418  62.360637    0.000000 -2.340219  

[2523 rows x 7 columns]

In [23]:
df.to_csv("Data/Prices/gridstatus_caiso_2023.csv")

## Solar + Wind Data (Emissions, Water Scarcity, Cost)

#### Capacity Factor

##### NSRDB

Reference: https://github.com/NREL/hsds-examples/tree/master

In [3]:
# takes around a minute to run 

solar = h5pyd.File(f"/nrel/nsrdb/v3/nsrdb_2013.h5")
meta = pd.DataFrame(solar['meta'][...])

In [4]:
# find the nearest timeseries
# code source: https://github.com/NREL/hsds-examples/blob/master/notebooks/03_NSRDB_introduction.ipynb
dset_coords = solar['coordinates'][...]
tree = cKDTree(dset_coords)

def nearest_site_solar(tree, lat_coord, lon_coord):
    """ 
    Find the nearest site in the solar dataset.

    Parameters
    ----------
        tree: scipy.spatial.cKDTree
            Lookup tree for nearest neighbor
        lat_coord: float
            Latitude
        lon_coord: float
            Longitude

    Returns
    -------
        Index of closest point in the solar dataset
    """

    lat_lon = np.array([lat_coord, lon_coord])
    _, pos = tree.query(lat_lon)
    return pos

In [ ]:
# collect DNI, DHI, GHI at every point
# takes around 4 min to run

dni_data = solar['dni']
dhi_data = solar['dhi']
ghi_data = solar['ghi']

dni_subbasins = np.zeros((dni_data.shape[0], huc8_ca.shape[0])) # rows: timestamps, # columns: locations
dhi_subbasins = np.zeros((dhi_data.shape[0], huc8_ca.shape[0])) # rows: timestamps, # columns: locations
ghi_subbasins = np.zeros((ghi_data.shape[0], huc8_ca.shape[0])) # rows: timestamps, # columns: locations

time_offset = np.zeros(huc8_ca.shape[0])

# for each California subbasin centroid, get the solar data of the closest point
for idx, (_, row) in enumerate(huc8_ca.iterrows()):
    centroid_idx = nearest_site_solar(tree, row['centroid_lat'], row['centroid_lon'])
    
    time_offset[idx] = meta.loc[centroid_idx, 'timezone']

    dni_tseries = dni_data[:, centroid_idx] / dni_data.attrs['psm_scale_factor']
    dhi_tseries = dhi_data[:, centroid_idx] / dhi_data.attrs['psm_scale_factor']
    ghi_tseries = ghi_data[:, centroid_idx] / ghi_data.attrs['psm_scale_factor'] # units: W/m^2. https://nsrdb.nrel.gov/data-sets/us-data

    dni_subbasins[:, idx] = dni_tseries
    dhi_subbasins[:, idx] = dhi_tseries
    ghi_subbasins[:, idx] = ghi_tseries

In [15]:
# by visual inspection of the data, we can tell that this time index data is UTC timezone. (California is -8 hours, according to the metadata)
# collect time index
time_index = pd.to_datetime(solar['time_index'][...].astype(str))

In [17]:
irradiation_xr = xr.Dataset()
irradiation_xr['DNI'] = xr.DataArray(dni_subbasins, dims=("Time", "HUC8"), coords={"Time": time_index, "HUC8": huc8_ca.index})
irradiation_xr['DHI'] = xr.DataArray(dhi_subbasins, dims=("Time", "HUC8"), coords={"Time": time_index, "HUC8": huc8_ca.index})
irradiation_xr['GHI'] = xr.DataArray(ghi_subbasins, dims=("Time", "HUC8"), coords={"Time": time_index, "HUC8": huc8_ca.index})

irradiation_xr.attrs['Timezone'] = time_offset

In [19]:
# save data
irradiation_xr.to_netcdf("Data/Solar/ca_irradiation.nc4")

# np.savez("Data/Solar/ca_irradiation.npz", dni=dni_subbasins, dhi=dhi_subbasins, ghi=ghi_subbasins, time_index=time_index)

##### WTK

Most common wind turbine models in California: V90-3.0 (489 count), GE1.5-77 (461 count), MM92 (245 count)

Most common wind turbine hub heights in California: 80.0m (1297 count), 65.0m (375 count), 24.0m (281 count)

Let's just fix the wind turbine model at GE1.5-77. Hub height 80.0m

https://en.wind-turbine-models.com/turbines/565-ge-vernova-ge-1.5s

https://nrel.github.io/turbine-models/DOE_GE_1.5MW_77.html

In [2]:
# question: what is typical turbine model, turbine height in California? Let's look at USWTDB: https://energy.usgs.gov/uswtdb/data/

existing_turbines_US = gpd.read_file("Data/Wind/uswtdbGeoJSON/uswtdb_V8_1_20250522.geojson")

# filter to California
existing_turbines_CA = existing_turbines_US[existing_turbines_US['t_state'] == "CA"]

existing_turbines_CA['t_model'].value_counts() # turbine models

existing_turbines_CA['t_hh'].value_counts() # turbine hub heights

t_hh
80.0     1297
65.0      375
24.0      281
60.0      206
78.5      161
24.6      132
55.0       95
67.0       85
50.0       74
91.5       64
84.0       60
85.0       55
94.0       51
64.0       50
69.0       45
68.5       42
40.0       36
92.0       32
82.0       28
81.5       23
90.0       23
89.0       22
105.0      19
63.0       19
45.0       17
87.0       13
39.5       13
30.0        8
56.5        6
Name: count, dtype: int64

Downloading wind speed data

In [3]:
wind = h5pyd.File("/nrel/wtk-us.h5")

wind_speed = wind['windspeed_80m'] # the most common wind speed is 80m among California wind turbines

# note: the wind data is hourly
wind_time = pd.DataFrame({'datetime': wind['datetime']})
wind_time['datetime'] = wind_time['datetime'].apply(dateutil.parser.parse)
wind_time['datetime'] = wind_time['datetime'].map(lambda x: x.replace(tzinfo=timezone.utc))

# get 2013 data when querying for specific locations, right now data is too large
# just select 2013 data
wind_time['year'] = wind_time['datetime'].map(lambda x: x.year)
wind_time = wind_time[wind_time['year'] == 2013]
indices_2013 = wind_time.index

In [6]:
# code source: https://github.com/NREL/hsds-examples/blob/master/notebooks/01_WTK_introduction.ipynb
def nearest_wind_site(data, lat_index, lon_index):
    """ 
    Get nearest site with wind data to input location.

    Parameters
    ----------
        data: h5pyd._hl.dataset.Dataset
            Full dataset, for coordinates
        lat_index: float
            Latitude
        lon_index: float
            Longitude

    Returns
    -------
        Latitude and longitude index of closest point in the wind dataset.
    """

    data_coords = data['coordinates']

    # find nearby point, convert to x/y indices
    projstring = """+proj=lcc +lat_1=30 +lat_2=60 
                    +lat_0=38.47240422490422 +lon_0=-96.0 
                    +x_0=0 +y_0=0 +ellps=sphere 
                    +units=m +no_defs """
    project_lcc = Proj(projstring)
    origin_ll = reversed(data_coords[0][0]) # origin. But flip order to get lon, lat
    origin = project_lcc(*origin_ll)

    coords = (lon_index, lat_index) # note the order of lon, lat
    coords = project_lcc(*coords)
    delta = np.subtract(coords, origin)
    ij = [int(round(x/2000)) for x in delta] # by rounding, we just get the nearest lattice point
    return tuple(reversed(ij))

In [17]:
# get wind speeds at closest location to each of the California HUC8 subbasins

# takes around 30 minutes to run

wind_speed_subbasins = np.zeros((wind_time.shape[0], huc8_ca.shape[0]))

# for each wind subbasin centroid, get wind data of closest point

while np.any(np.sum(wind_speed_subbasins, axis=0) == 0):
    for idx, (_, row) in enumerate(huc8_ca.iterrows()):
        if np.sum(wind_speed_subbasins[:,idx]) == 0:
            try:
                centroid_idx = nearest_wind_site(wind, row['centroid_lat'], row['centroid_lon'])

                print(centroid_idx)

                
                wind_tseries = wind_speed[indices_2013, centroid_idx[0], centroid_idx[1]] # units: m/s.
                
                time.sleep(30)
                wind_tseries = wind_speed[indices_2013, centroid_idx[0], centroid_idx[1]] # units: m/s.

                wind_speed_subbasins[:, idx] = wind_tseries
            except:
                pass
    
    time.sleep(10)

(593, 635)
(608, 546)
(613, 565)
(618, 598)
(622, 534)
(632, 565)
(638, 641)
(642, 535)
(655, 538)
(662, 515)
(666, 568)
(669, 542)
(681, 608)
(688, 530)
(685, 658)
(723, 577)
(717, 648)
(745, 557)
(753, 628)
(781, 542)
(674, 506)
(681, 499)
(695, 503)
(705, 476)
(707, 492)
(718, 464)
(733, 451)
(731, 476)
(743, 514)
(767, 481)
(793, 561)
(794, 521)
(788, 587)
(785, 603)
(814, 516)
(829, 537)
(848, 553)
(882, 531)
(884, 550)
(912, 522)
(791, 482)
(803, 482)
(819, 493)
(817, 454)
(836, 491)
(855, 454)
(857, 510)
(883, 465)
(883, 501)
(908, 478)


ERROR:root:POST error: 503
ERROR:root:POST error - status_code: 503, reason: Service Unavailable


(923, 477)
(938, 470)
(943, 506)
(958, 499)
(975, 470)
(980, 492)
(990, 472)
(1002, 494)
(1026, 479)
(1043, 483)
(931, 514)
(1065, 518)
(1084, 525)
(1112, 531)
(1132, 539)
(1057, 499)
(1081, 510)
(1109, 502)
(1131, 525)
(1138, 492)
(658, 462)
(718, 417)
(746, 421)
(752, 425)
(767, 423)
(766, 439)
(780, 442)
(807, 427)
(863, 422)
(887, 445)
(766, 409)
(827, 393)
(839, 404)
(883, 405)
(895, 385)
(914, 392)
(921, 373)
(931, 394)
(959, 398)
(971, 367)
(914, 426)
(934, 440)
(940, 419)
(945, 453)
(956, 456)
(966, 453)
(980, 438)
(989, 437)
(999, 451)
(1012, 436)
(971, 384)
(972, 417)
(992, 395)
(1005, 358)
(1007, 372)
(1013, 391)
(1018, 414)
(1042, 401)
(1044, 380)
(1039, 357)
(1013, 466)
(1032, 435)
(1052, 462)
(1053, 436)
(1063, 424)
(1071, 426)
(1076, 449)
(1089, 444)
(1092, 429)
(1100, 466)
(1060, 386)
(1079, 363)
(1081, 411)
(1092, 372)
(1099, 353)
(1100, 391)
(1118, 407)
(1122, 374)
(1138, 409)
(1145, 425)
(1116, 436)
(1119, 451)
(1143, 444)
(1144, 463)
(1161, 438)
(1156, 397)
(1173, 3

In [23]:
# save wind speeds to csv

wind_speed_subbasins_df = pd.DataFrame(wind_speed_subbasins, index=wind_time['datetime'], columns=huc8_ca.index)
wind_speed_subbasins_df.to_csv("Data/Wind/ca_wind_speed.csv")